File: dailyQuestion/2020-06/06-26/python/solution_dp.py
def wordBreak(s: str, wordDict: [str]) -> bool:
    wordDictSet = {}
    maxlen = 0
    for item in wordDict:
        wordDictSet[item] = True
        if len(item) > maxlen:
            maxlen = len(item)
    length = len(s)
    dp = [False] * (length + 1)
    dp[0] = True
    for i in range(1, length + 1):
        for j in range(0, i):
            if dp[j] and wordDictSet.get(s[j:i]):
                dp[i] = True
                break
    return dp[length]

if __name__ == "__main__" :
    s = "catsandog"
    wordDict = ["cats", "dog", "sand", "and", "cat"]
    result = wordBreak(s, wordDict)
    print(result)

File: dailyQuestion/2020-06/06-26/golang/solution_dp.go
package main

import (
	"fmt"
)

func wordBreak(s string, wordDict []string) bool {
	wordDictSet := make(map[string]bool)
	maxlen := 0
	for _, item := range wordDict {
		wordDictSet[item] = true
		if len(item) > maxlen {
			maxlen = len(item)
		}
	}
	length := len(s)
	dp := make([]bool, length+1)
	dp[0] = true
	for i := 1; i <= length; i++ {
		for j := 0; j < i; j++ {
			if len(s[j:i]) > maxlen {
				continue
			}
			if dp[j] && wordDictSet[s[j:i]] {
				dp[i] = true
				break
			}
		}
	}
	return dp[length]
}

func main() {
	s := "catsand"
	wordDict := []string{"cats", "dog", "sand", "and", "cat"}
	result := wordBreak(s, wordDict)
	fmt.Println(result)
}
